### PR TITLE
feat: implement function `split_part`

### PR DESCRIPTION
**Pull request by alissa-tung**
## What's changed and what's your intention?

implement `split_part`

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests

## Refer to a related PR or issue link (optional)

#2791
meta #112

### CODE REVIEW

**Comment by alissa-tung** [comment]
still draft

File: src/expr/src/expr/expr_split_part.rs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+        let row_len = input.cardinality();
+        let mut builder = Utf8ArrayBuilder::new(row_len)?;
+
+        for i in 0..row_len {
+            let string_column = string_columns.value_at(i);
+            let delimiter_column = delimiter_columns.value_at(i);
+            let nth_column = nth_columns.value_at(i);
+
+            // FIXME
```
Comment (alissa-tung) [inline src/expr/src/expr/expr_split_part.rs:59]: Looking for suggestions on:
- what error handling method to use in order to deal with the below three unwrapping
- what should be as expected behavior, when index is zero

File: src/frontend/src/expr/function_call.rs [hunk 2]
```diff
@@ -135,6 +139,31 @@ impl FunctionCall {
...
+                }
+
+                inputs = inputs
+                    .into_iter()
+                    .enumerate()
+                    .map(|(i, input)| match i {
+                        0 => input.cast_implicit(DataType::Varchar),
+                        1 => input.cast_implicit(DataType::Varchar),
+                        2 => input.cast_implicit(DataType::Int64),
+                        _ => todo!(),
```
Comment (alissa-tung) [inline src/frontend/src/expr/function_call.rs:161]: Since we had check length, what should be insert to fill to clause?

**Comment by alissa-tung** [comment]
cc @TennyZhuang @skyzh

File: rust-toolchain [hunk 3]
```diff
@@ -1 +1 @@
-nightly-2022-05-24
+nightly-2022-05-25
```
Comment (xiangjinwu) [inline rust-toolchain:1]: If a toolchain update is really needed, send it as a standalone PR.

File: src/common/src/array/mod.rs [hunk 4]
```diff
@@ -125,10 +125,10 @@ pub trait ArrayBuilder: Send + Sync + Sized + 'static {
 /// `Array` must be built with an `ArrayBuilder`. The array trait provides several
 /// unified interface on an array, like `len`, `value_at` and `iter`.
 ///
-/// The `Builder` associated type is the builder for this array. The `Iter` associated
-/// type is the iterator of this array. And the `RefItem` is the item you could
-/// retrieve from this array.
+/// The `Builder` associated type is the builder for this array.
 ///
+/// The `Iter` associated type is the iterator of this array. And the `RefItem` is
+/// the item you could retrieve from this array.
```
Comment (xiangjinwu) [inline src/common/src/array/mod.rs:131]: Intended change or accidental?

File: src/expr/src/expr/expr_split_part.rs [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+#[derive(Debug)]
+pub struct SplitPartExpression {
+    return_type: DataType,
+    string_expr: BoxedExpression,
+    delimiter_expr: BoxedExpression,
+    nth_expr: BoxedExpression,
+}
+
+impl Expression for SplitPartExpression {
```
Comment (xiangjinwu) [inline src/expr/src/expr/expr_split_part.rs:36]: Avoid `impl Expression` directly when possible. Take a look at `TernaryBytesExpression`.

File: src/expr/src/expr/expr_split_part.rs (hunk 1 above)
Comment (xiangjinwu) [reply src/expr/src/expr/expr_split_part.rs:59]: * The three unwrapping corresponding to `null`s at each position.
* As for expected behavior when an argument is `null` or when index is zero or when index out of bound, just try it in PostgreSQL.

File: src/expr/src/expr/expr_split_part.rs [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+    fn eval(&self, input: &DataChunk) -> Result<ArrayRef> {
+        let string_columns = self.string_expr.eval(input)?;
+        let string_columns = string_columns.as_utf8();
+
+        let delimiter_columns = self.delimiter_expr.eval(input)?;
+        let delimiter_columns = delimiter_columns.as_utf8();
+
+        let nth_columns = self.nth_expr.eval(input)?;
+        let nth_columns = nth_columns.as_int64();
```
Comment (alissa-tung) [inline src/expr/src/expr/expr_split_part.rs:49]: Here `nth` is treated as `i64`, what numerical types should we support here?

File: src/frontend/src/expr/function_call.rs (hunk 2 above)
Comment (xiangjinwu) [reply src/frontend/src/expr/function_call.rs:161]: Given this function has a fixed signature: `(varchar, varchar, i32) -> varchar`, add it in `type_inference::build_type_derive_map` rather than manually checking here.

Functions listed here are all variadic. Maybe you can do me a favor and add this as comment in this function.

File: src/frontend/test_runner/tests/testdata/expr.yaml [hunk 7]
```diff
@@ -61,7 +61,7 @@
 - sql: |
     /* in-list with misaligned types */
     SELECT true in (3, 1.0, 2);
-  binder_error: 'Bind error: types Boolean and Int32 cannot be matched'
+  binder_error: "Bind error: types Boolean and Int32 cannot be matched"
```
Comment (xiangjinwu) [inline src/frontend/test_runner/tests/testdata/expr.yaml:64]: Are you using a different version of `yq`?

File: src/expr/src/expr/expr_split_part.rs (hunk 6 above)
Comment (xiangjinwu) [reply src/expr/src/expr/expr_split_part.rs:49]: It is `integer` (`int4` / `int32`) in PostgreSQL:
https://github.com/postgres/postgres/blob/de89d8711e2b70150a2c5b3215d99da108fbe1e8/src/include/catalog/pg_proc.dat#L3638-L3640
```
test=# select split_part(text '', text '', 3::bigint);
ERROR:  function split_part(text, text, bigint) does not exist
LINE 1: select split_part(text '', text '', 3::bigint);
               ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
```

File: src/common/src/array/mod.rs (hunk 4 above)
Comment (alissa-tung) [reply src/common/src/array/mod.rs:131]: As intended, since `Iter` is the valid way to consume arrays. I thought it would be better to put `Iter` and `RefItem` in the same block, instead of `Builder` and `Iter` on the same line, especially for quote `Iter` and EOL.

File: src/frontend/src/expr/function_call.rs (hunk 2 above)
Comment (alissa-tung) [reply src/frontend/src/expr/function_call.rs]: thx. [41909ae](https://github.com/singularity-data/risingwave/pull/2822/commits/41909ae9abf146401014e7d4cc8f5ed40b9333fa)

File: rust-toolchain (hunk 3 above)
Comment (alissa-tung) [reply rust-toolchain]: thx [825eaec](https://github.com/singularity-data/risingwave/pull/2822/commits/825eaeca58052deb9b9ade74595ba551b5c0ea2e)

File: src/expr/src/expr/expr_split_part.rs (hunk 6 above)
Comment (alissa-tung) [reply src/expr/src/expr/expr_split_part.rs]: thx, fixed

File: src/frontend/test_runner/tests/testdata/expr.yaml (hunk 7 above)
Comment (alissa-tung) [reply src/frontend/test_runner/tests/testdata/expr.yaml:64]: I do not know this repo use `yq` before. After a grep I only found [this](https://github.com/singularity-data/risingwave/blob/606925831c462a99796fccf865adba8a2e91a656/docs/developer-guide.md?plain=1#L326).
Is there any docs about which args should be use with `yq` when formating `expr.yaml` or if there exists some scripts to do this stuff?

File: src/expr/src/expr/expr_split_part.rs (hunk 5 above)
Comment (alissa-tung) [reply src/expr/src/expr/expr_split_part.rs]: thx, fixed

is [455d9d0](https://github.com/singularity-data/risingwave/pull/2822/commits/455d9d0aa399079e9fc61d1d6ddd08c9eaaf7fe1) okay to return error code here?

File: src/expr/src/expr/expr_split_part.rs (hunk 1 above)
Comment (alissa-tung) [reply src/expr/src/expr/expr_split_part.rs]: thx, fixed [dcc5b6a](https://github.com/singularity-data/risingwave/pull/2822/commits/dcc5b6a225a62d64fd5fc478bb0e741509e01899)

File: src/expr/src/vector_op/split_part.rs [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+        //           if first or last field, return input string, else empty string
+        if nth_expr == 1 || nth_expr == -1 {
+            string_expr
+        } else {
+            Default::default()
+        }
+    } else {
+        match nth_expr.cmp(&0) {
+            std::cmp::Ordering::Equal => {
+                return Err(RwError::from(ErrorCode::InvalidParameterValue(
```
Comment (alissa-tung) [inline src/expr/src/vector_op/split_part.rs:40]: @xiangjinwu is it okay to return error here?

File: src/expr/src/vector_op/split_part.rs [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+    nth_expr: i32,
+    writer: BytesWriter,
+) -> Result<BytesGuard> {
+    let mut split = string_expr.split(delimiter_expr);
+    let nth_val = if string_expr.is_empty() {
+        // postgres: return empty string for empty input string
+        Default::default()
+    } else if delimiter_expr.is_empty() {
+        // postgres: handle empty field separator
+        //           if first or last field, return input string, else empty string
```
Comment (xiangjinwu) [inline src/expr/src/vector_op/split_part.rs:31]: * It seems the case of empty `string_expr` can be handled naturally by logic below without special treatment.
* As for empty `delimiter_expr`, the `split` iterator can be replaced by `std::iter::once(string_expr)`, but the type checker is unhappy unless we box it or use a generic helper function on `impl Iterator<Item=&str>`. So I am okay with current implementation, except that nth == 0 should return error.

File: src/expr/src/vector_op/split_part.rs [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    use risingwave_common::error::{ErrorCode, Result, RwError};
+
+    use super::split_part;
+
+    #[test]
+    fn test_split_part() {
+        let cases: Vec<(&str, &str, i32, Result<&str>)> = vec![
+            ("", "@", 1, Ok("")),
+            ("", "@", -1, Ok("")),
+            ("joeuser@mydatabase", "", 1, Ok("joeuser@mydatabase")),
```
Comment (xiangjinwu) [inline src/expr/src/vector_op/split_part.rs:72]: Can we add a case `("@@@", "@@", -1, Ok("@"))`? This makes sure `rsplit` is not used internally when `nth` is negative.

File: src/frontend/test_runner/tests/testdata/expr.yaml (hunk 7 above)
Comment (xiangjinwu) [reply src/frontend/test_runner/tests/testdata/expr.yaml:64]: This file is generated by `./risedev do-apply-planner-test` and the `yq` args should be here:
https://github.com/singularity-data/risingwave/blob/a4d85ede4dd96058668c1fe1a38a4afe5473e879/src/frontend/test_runner/planner_test.toml#L13

**Review by xiangjinwu — COMMENTED** [review]
Nice work! Just some minor things:

File: src/frontend/test_runner/tests/testdata/expr.yaml (hunk 7 above)
Comment (alissa-tung) [reply src/frontend/test_runner/tests/testdata/expr.yaml:64]: Thx.

Would it be better to add this line in [the task](https://github.com/singularity-data/risingwave/blob/a4d85ede4dd96058668c1fe1a38a4afe5473e879/src/frontend/test_runner/planner_test.toml#L49)?

If so I gonna file another merge request for this

```sh
    sed -i '1i# This file is generated and formatted by running the task `./risedev do-apply-planner-test`.\n' "$f"
```

File: src/expr/src/vector_op/split_part.rs [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+                split
+                    .get(nth_expr)
+                    .map_or(Default::default(), <&str>::clone)
+            }
+        }
+    };
+
+    let mut writer = writer.begin();
+    writer.write_ref(nth_val).unwrap();
+    Ok(writer.finish().unwrap())
```
[the remark this answers is not in the document]
Comment (alissa-tung) [reply src/expr/src/vector_op/split_part.rs]: thx. is [99aefb0](https://github.com/singularity-data/risingwave/pull/2822/commits/99aefb02500cf7153028eb9b4aa1592409b9ba86) as expected?

File: src/expr/src/vector_op/split_part.rs [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+                    "field position must not be zero".into(),
+                )));
+            }
+            std::cmp::Ordering::Greater => split.nth(nth_expr as usize - 1).unwrap_or_default(),
+            std::cmp::Ordering::Less => {
+                let split = split.collect::<Vec<_>>();
+                let nth_expr = (split.len() as i32 + nth_expr) as usize;
+                split
+                    .get(nth_expr)
+                    .map_or(Default::default(), <&str>::clone)
```
[the remark this answers is not in the document]
Comment (alissa-tung) [reply src/expr/src/vector_op/split_part.rs]: [57e54c8](https://github.com/singularity-data/risingwave/pull/2822/commits/57e54c83bfc078fe954d5d81c0e1817557b9527f)

File: src/expr/src/vector_op/split_part.rs (hunk 9 above)
Comment (alissa-tung) [reply src/expr/src/vector_op/split_part.rs:37]: thx!
[a899543](https://github.com/singularity-data/risingwave/pull/2822/commits/a899543286d4e1eaa182d08f036d0f77bcfcadc8)

File: src/expr/src/vector_op/split_part.rs [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+        match nth_expr.cmp(&0) {
+            std::cmp::Ordering::Equal => {
+                return Err(RwError::from(ErrorCode::InvalidParameterValue(
+                    "field position must not be zero".into(),
+                )));
+            }
+            std::cmp::Ordering::Greater => split.nth(nth_expr as usize - 1).unwrap_or_default(),
+            std::cmp::Ordering::Less => {
+                let split = split.collect::<Vec<_>>();
+                let nth_expr = (split.len() as i32 + nth_expr) as usize;
```
[the remark this answers is not in the document]
Comment (alissa-tung) [reply src/expr/src/vector_op/split_part.rs]: thx

[da0ba96](https://github.com/singularity-data/risingwave/pull/2822/commits/da0ba968393cbf49f74ec904f9c4602c8eb3d8db)

File: src/expr/src/vector_op/split_part.rs (hunk 10 above)
Comment (alissa-tung) [reply src/expr/src/vector_op/split_part.rs:84]: [da0ba96](https://github.com/singularity-data/risingwave/pull/2822/commits/da0ba968393cbf49f74ec904f9c4602c8eb3d8db)

**Comment by xiangjinwu** [comment]
https://github.com/singularity-data/risingwave/runs/6621101096?check_suite_focus=true#step:10:65
```
error: unresolved link to `crate::expr::type_inference::build_type_derive_map`
  --> src/frontend/src/expr/function_call.rs:93:21
   |
93 |     /// [elsewhere](crate::expr::type_inference::build_type_derive_map).
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no item named `type_inference` in module `expr`
   |
note: the lint level is defined here
  --> src/frontend/src/lib.rs:26:9
   |
26 | #![deny(rustdoc::broken_intra_doc_links)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: could not document `risingwave_frontend`
```

**Comment by alissa-tung** [comment]
> https://github.com/singularity-data/risingwave/runs/6621101096?check_suite_focus=true#step:10:65
> 
> ```
> error: unresolved link to `crate::expr::type_inference::build_type_derive_map`
>   --> src/frontend/src/expr/function_call.rs:93:21
>    |
> 93 |     /// [elsewhere](crate::expr::type_inference::build_type_derive_map).
>    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no item named `type_inference` in module `expr`
>    |
> note: the lint level is defined here
>   --> src/frontend/src/lib.rs:26:9
>    |
> 26 | #![deny(rustdoc::broken_intra_doc_links)]
>    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> error: could not document `risingwave_frontend`
> ```



> https://github.com/singularity-data/risingwave/runs/6621101096?check_suite_focus=true#step:10:65
> 
> ```
> error: unresolved link to `crate::expr::type_inference::build_type_derive_map`
>   --> src/frontend/src/expr/function_call.rs:93:21
>    |
> 93 |     /// [elsewhere](crate::expr::type_inference::build_type_derive_map).
>    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no item named `type_inference` in module `expr`
>    |
> note: the lint level is defined here
>   --> src/frontend/src/lib.rs:26:9
>    |
> 26 | #![deny(rustdoc::broken_intra_doc_links)]
>    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> error: could not document `risingwave_frontend`
> ```

@xiangjinwu 

I thought it is caused by visibility.
[c3ce096](https://github.com/singularity-data/risingwave/pull/2822/commits/c3ce096024f28dcfdaf0ea3314f5f83ad1a6757b)
workaround

**Comment by xiangjinwu** [comment]
> I thought it is caused by visibility. [c3ce096](https://github.com/singularity-data/risingwave/pull/2822/commits/c3ce096024f28dcfdaf0ea3314f5f83ad1a6757b) workaround

Yes it is due to visibility. But we should not make things `pub` simply for references in docs. As the references are for developers updating the code rather than calling the code, maybe they should not be part of public doc. (i.e. use `//` rather than `///` for parts explaining `type inference` vs `FunctionCall new`)

**Comment by alissa-tung** [comment]
> > I thought it is caused by visibility. [c3ce096](https://github.com/singularity-data/risingwave/pull/2822/commits/c3ce096024f28dcfdaf0ea3314f5f83ad1a6757b) workaround
> 
> Yes it is due to visibility. But we should not make things `pub` simply for references in docs. As the references are for developers updating the code rather than calling the code, maybe they should not be part of public doc. (i.e. use `//` rather than `///` for parts explaining `type inference` vs `FunctionCall new`)

ok. I am going to revert that commit and use double for now

**Comment by alissa-tung** [comment]
cc @xiangjinwu done

**Review by xiangjinwu — APPROVED** [review]
Thank you for your contribution!
